### DISCUSSION TITLE
Random unit vector optimization

### DISCUSSION THREAD

**Discussion by DeveloprOfThings** [RayTracing/raytracing.github.io#941] [Q&A] [answered]
In section 8.5 of Ray Tracking in One Weekend, to achieve "True Lambertian Reflection" the function `random_in_unit_sphere()` is transformed into a unit vector. Since this returned vec3 is always normalized, would it be more efficient and still result in a similar image to remove the looping from the algorithm and just normalize a vector returned from the function `Random(-1.d, 1.d);` in the `vec3` class?

I have run the code myself and the results look incredibly similar but not identical (from my understanding the images will never be identical; even with subsequent runs). If my intuition about the two algorithms leading to accurate images, could anyone explain where I've gone wrong?

**Comment by hollasch** [comment, marked as answer]
If you use the `vec3::random(double,double)` function, you're essentially generating random points in the cube with vertices at (-1,-1,-1) and (+1,+1,+1). However, the diameter from the center of one side to the opposite side is smaller than the diameter from one corner to the opposite corner. In fact, it's ≅ 1.7 times (√3) longer, so more points are generated along this diameter than along the center-to-center diameter.

On normalization, these points are "squished" (projected) to the point on the surface of the embedded sphere. As a result, you get an uneven distribution of points on the surface of this sphere.

In contrast, the rejection method presented in the book ensures that only points within the sphere are accepted, and such points have no bias toward any particular region of the sphere.
